crates/interpreter/src/frame.rs

```diff
@@ -0,0 +1,176 @@
+            local_values[*v] = literal_value
+        }
+        let alloca_region = Vec::new();
+
+        Self {
+            ret_addr,
+            local_values,
+            alloca_region,
+        }
+    }
+
+    pub fn load(&mut self, /*ctx: Context,*/ v: Value, dfg: &DataFlowGraph) -> Literal {
+        if !self.is_assigned(v) {
+            let v = match dfg.value_data(v) {
+                ValueData::Insn { insn, .. } => {
+                    let result_v = dfg.insn_result(*insn).unwrap();
+                    if self.is_assigned(result_v) {
+                        return self.local_values[result_v];
+                    }
+                    result_v
+                }
+                _ => v,
+            };
+            let i256 = dfg.value_imm(v).unwrap().as_i256();
+            self.local_values[v] = Literal(i256);
+        }
+        self.local_values[v]
+    }
+
+    pub fn map(&mut self, literal: Literal, insn: Insn, dfg: &DataFlowGraph) {
+        let v = dfg.insn_result(insn).unwrap();
+        debug_assert!(!self.is_assigned(v));
+        self.local_values[v] = literal
+    }
+
+    pub fn alloca(&mut self, ctx: &ModuleCtx, ty: Type, insn: Insn, dfg: &DataFlowGraph) {
+        let v = dfg.insn_result(insn).unwrap();
+        debug_assert!(!self.is_assigned(v));
+
+        let addr = self.alloca_region.len();
+
+        let size_of_data = byte_size_of_ty(ctx, ty);
+        for _ in 0..size_of_data {
+            self.alloca_region.push(0u8);
+        }
+        self.local_values[v] = Literal::from_usize(addr);
+    }
+
+    pub fn gep(&mut self, ctx: &ModuleCtx, args: &[Value], dfg: &DataFlowGraph) -> Literal {
+        let ptr_v = args[0];
+        let ptr = self.load(ptr_v, dfg);
+        let base_addr = ptr.as_usize();
+        let ptr_ty = dfg.value_ty(ptr_v);
+        debug_assert!(ctx.with_ty_store(|s| s.is_ptr(ptr_ty)));
+
+        let pointee_ty = ctx.with_ty_store(|s| s.deref(ptr_ty)).unwrap();
+        debug_assert!(!pointee_ty.is_integral() && !ctx.with_ty_store(|s| s.is_ptr(ptr_ty)));
+
+        let mut cmpd_ty = to_cmpd_ty(pointee_ty);
+        let mut offset = 0usize;
+        for arg in &args[1..] {
+            let index = self.load(*arg, dfg).as_usize();
+
+            ctx.with_ty_store(|s| match s.resolve_compound(cmpd_ty.unwrap()) {
+                CompoundTypeData::Array { elem, .. } => {
+                    offset += index * byte_size_of_ty(ctx, *elem);
+                    cmpd_ty = to_cmpd_ty(*elem);
+                }
+                CompoundTypeData::Struct(data) => {
+                    for ty in &data.fields[..index] {
+                        offset += byte_size_of_ty(ctx, *ty);
+                    }
+                    cmpd_ty = to_cmpd_ty(data.fields[index]);
+                }
+                _ => unreachable!(),
+            })
+        }
+        Literal::from_usize(base_addr + offset)
+    }
+
+    pub fn ldr(&mut self, ctx: &ModuleCtx, ptr: Value, insn: Insn, dfg: &DataFlowGraph) {
```

Comment:
Ditto.
It'd be nicer just to load the literal from the address specified the `ptr`.